packages/block-library/src/image/block.json

```diff
@@ -93,7 +93,6 @@
 			"color": true,
 			"radius": true,
 			"width": true,
-			"__experimentalSelector": "img, .wp-block-image__crop-area",
```

Comment:
I understand that we added `.wp-block-image__crop-area` styles at https://github.com/WordPress/gutenberg/pull/31366 because we wanted to retain the border styles in the editor when the user is cropping the image. 

I ran Gutenberg before this PR to see how it works and this is what I see:

- cropping from default retains the styles

[Gravação de ecrã a partir de 27-03-2023 20:10:14.webm](https://user-images.githubusercontent.com/583546/228030161-f95cd2b6-77b9-4ed1-bec0-de805aaf6097.webm)

- cropping when the rounded style is applied does not retain anything

[Gravação de ecrã a partir de 27-03-2023 20:09:18.webm](https://user-images.githubusercontent.com/583546/228030176-69240355-6481-417e-ae9a-96c9004b3167.webm)

This is the only use case I could see for the `editorStyles` API. Is that correct? Is there any other use case I'm unaware of? Personally, I don't see a pressing need for `editorSelectors` if it is just this single case. I wouldn't have it. An editor API is something we need to add and support forever and comes with a cost. Instead, what I feel we should do is:

- Removing the crop class from the selectors, so it's not part of the front-end output unnecessarily.
- Making the crop tool be aware of block supports by taking the data applied to the block and re-apply it to its own wrapper (block supports, block styles, etc.). Judging by the current state of the "round view" (it doesn't work), it doesn't seem too pressing for the crop operation to reflect exactly the styles in "normal" view.

Reply:
Thanks for digging into this @oandregal 🙇 

> cropping when the rounded style is applied does not retain anything

At this stage, the rounded Block Style doesn't get applied to the cropping area as noted. Given how it may impact the resulting crop, it would make sense to tweak the CSS for that to apply to the cropping area as well. 

I can make that tweak once the selectors API is ironed out (if we are still applying other borders to the cropping area).

> This is the only use case I could see for the editorStyles API. Is that correct? Is there any other use case I'm unaware of? 

The Post Featured Image block applies the border to the image placeholder which is also editor only. The driver here was for those borders to be shown for things like the Query block in the Site Editor. There is also an issue to [apply borders to the Image block placeholders](https://github.com/WordPress/gutenberg/issues/49144) when not selected in the editors. I'd expect this to be happening for placeholder related blocks with border support.

At present, our blocks using `ServerSideRender` end up with a wrapping div with the same block classname. Global Styles for the block can then be applied twice, which is fine for say typography, but padding or margin on multiple elements creates a real spacing issue. Editor-only selectors here could be used to apply the global styles to only the desired element. That said, the use of `ServerSideRender` is discouraged and mainly for legacy purposes so not a strong use case pointing to the need for editor-only selectors.

The last use-case off the top of my head would be blocks that use editor features like resizable boxes (Cover), that introduce additional markup in the editor. That resizable box has been a blocker to getting borders onto the Cover block. Being able to change the target for Global Styles on that block in the editor per feature would have provided more options there. At present, there is a PR having to try and move that resizable box to a block popover but that comes with other trade-offs and might not be accepted, leaving us with a possible need for more flexibility with Global Styles in the editor.

> Personally, I don't see a pressing need for editorSelectors if it is just this single case. I wouldn't have it. 

I'm open to the idea of omitting the editor-only selectors functionality, even if it is just a temporary removal so we can land the main selectors API and get that into use. An editor API is something we need to add and support forever and comes with a cost.

> Making the crop tool be aware of block supports by taking the data applied to the block and re-apply it to its own wrapper (block supports, block styles, etc.).

I believe we now have access to merged Global Styles data in the block editor as well as the site editor so this should be feasible.

I'm not sure I see this working for all use cases though. It does when we are looking to apply styles to additional elements e.g. the image cropping area, but it could be more hit and miss when we need to completely change or override the target of the styles in the editor.

Let me know what you think about the other potential use cases and whether you're still leaning towards removing the editor-only selectors.

Reply:
After talking through some of the issues here privately with @talldan, I'd like to propose the following plan:

1. Create a new PR that addresses:
    - some of the gaps of the selectors API (e.g. useGlobalStylesOutput), 
    - fixes the global style variations with feature selectors, 
    - ensures duotone selectors are under `filter.duotone`
    - and **removes the editor-only selectors functionality** 
2. If there is any pushback or further cause for delay, revert the selectors API via https://github.com/WordPress/gutenberg/pull/49372
3. Roll back the addition of the `editorSelectors` block type property in the related [core patch](https://github.com/WordPress/wordpress-develop/pull/3942)
4. Update or close any of the existing related Duotone PRs as needed

Any thoughts or concerns with this approach?

Reply:
The plan sounds good to me.

Thanks for sharing the additional use cases for the editor selectors API. Those are things worth looking at. I don't have a concrete fix now, though I can share some rough notes to explore later (not completely sure of trade-offs):

- Can the block edit function absorb that complexity? The selectors API only contains front-end selectors, but we provide an editor hook, so the blocks can update its selectors depending on the "context" (if the block is in a special action such as cropping, etc.)? This depends on the global styles being created in the post editor instead of pre-created in the server and enqueued, which seems a direction people increasingly favor anyway.
- Can we still do what we do today (add editor-only selectors to the selectors API) but provide a filter to edit the block selectors?

Reply:
> Can the block edit function absorb that complexity? The selectors API only contains front-end selectors, but we provide an editor hook, so the blocks can update its selectors depending on the "context" (if the block is in a special action such as cropping, etc.)?

My opinion is that the `editorSelectors` next to `selectors` would be a more "natural" API for consumers since we already have `editorScripts` and `editorStyles`. I don't think it would be as intuitive to have a hook that modifies the selectors when you're in the editor.

Reply:
> Can the block edit function absorb that complexity?

Personally, I feel like leveraging hooks or filters here would add to the complexity and obscurity for editor selectors.

> The selectors API only contains front-end selectors, but we provide an editor hook, so the blocks can update its selectors depending on the "context"

Maybe I'm not quite clear on how this would work but I'm not 100% sold on the use of hooks here. We'd need to be offering those blocks a means of forcing the regeneration of theme.json block metadata, sounds like we could be losing a lot of the theme.json performance gains you've worked so hard to achieve.

> Can we still do what we do today (add editor-only selectors to the selectors API) but provide a filter to edit the block selectors?

So I understand this proposal, we would:
1. Keep `editorSelectors` in the block.json
2. Apply filters in `WP_Theme_JSON::get_blocks_metadata()`
3. Add a filter somewhere, once we've established we're in the editor, to generate metadata taking into account the editor selectors

I think the main thing I haven't got my head around is preventing generating the metadata unnecessarily.

> My opinion is that the editorSelectors next to selectors would be a more "natural" API for consumers since we already have editorScript and editorStyle.

Agreed. I think that `editorSelectors` would be more natural for consumers and more consistent with existing APIs.

Which in terms of the hook/filter proposals above, leans towards the latter of the two.